core/templates/pages/teach-page/teach-page.component.spec.ts

```diff
@@ -15,112 +15,193 @@
-    setTimeout(function() {
-      expect(windowRef.nativeWindow.location.hash).toBe('#teach');
-      expect(ctrl.activeTabName).toBe('teach');
-      done();
-    });
+    expect(windowRef.nativeWindow.location.hash).toBe('#teach');
+    expect(component.activeTabName).toBe('teach');
   });
 
-  it('should click on participation tab', function(done) {
-    ctrl.$onInit();
-    expect(ctrl.activeTabName).toBe('teach');
-
-    ctrl.onTabClick('participation');
+  it('should click on participation tab', (() => {
+    component.ngOnInit();
+    expect(component.activeTabName).toBe('teach');
 
-    // Function setTimeout is being used here in order to wait onhashchange
-    // event to finish. setTimeout is executed only after call stack is empty.
-    // Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop
-    setTimeout(function() {
-      expect(windowRef.nativeWindow.location.hash).toBe('#participation');
-      expect(ctrl.activeTabName).toBe('participation');
-      done();
-    });
-  });
+    component.onTabClick('participation');
+    expect(windowRef.nativeWindow.location.hash).toBe('#participation');
+    expect(component.activeTabName).toBe('participation');
+  }));
 
-  it('should activate teach tab on init', function() {
+  it('should activate teach tab on init', () => {
     windowRef.nativeWindow.location.hash = '#teach';
 
-    ctrl.$onInit();
+    component.ngOnInit();
 
     expect(windowRef.nativeWindow.location.hash).toBe('#teach');
-    expect(ctrl.activeTabName).toBe('teach');
+    expect(component.activeTabName).toBe('teach');
   });
 
-  it('should activate participation tab on init', function() {
+  it('should activate participation tab on init', () => {
     windowRef.nativeWindow.location.hash = '#participation';
 
-    ctrl.$onInit();
+    component.ngOnInit();
 
     expect(windowRef.nativeWindow.location.hash).toBe('#participation');
-    expect(ctrl.activeTabName).toBe('participation');
+    expect(component.activeTabName).toBe('participation');
   });
 
   it('should get static image url', function() {
-    expect(ctrl.getStaticImageUrl('/path/to/image')).toBe(
+    expect(component.getStaticImageUrl('/path/to/image')).toBe(
       '/assets/images/path/to/image');
   });
 
-  it('should apply to teach with oppia', function() {
-    var applyToTeachWithOppiaEventSpy = spyOn(
-      SiteAnalyticsService, 'registerApplyToTeachWithOppiaEvent')
+  it('should apply to teach with oppia', fakeAsync(() => {
+    const applyToTeachWithOppiaEventSpy = spyOn(
+      sas, 'registerApplyToTeachWithOppiaEvent')
       .and.callThrough();
 
-    ctrl.$onInit();
+    component.ngOnInit();
     spyOnProperty(windowRef, 'nativeWindow').and.returnValue({
-      location: ''
+      location: {
+        href: ''
+      }
     });
-    ctrl.onApplyToTeachWithOppia('/login');
-    $timeout.flush(150);
+    component.onApplyToTeachWithOppia();
+
+    tick(200);
```

Comment:
Is this really needed?

Reply:
So, I needed to simulate the time before the Macrotasks (setTimeout) passed, and so used tick

Reply:
For reference: https://blog.bitsrc.io/microtask-and-macrotask-a-hands-on-approach-5d77050e2168

Reply:
Use fixture.detectChanges(). It might work

Reply:
Okay, so using (done) and calling it inside the timeout after the expect statements worked. I saw this on donate-page.component, which used a similar third party link. But what exactly does it do? I'll add fixture.detect changes as well

Reply:
Removed done for detectchanges.
detect changes worked, but I had to increase the timeout delay to avoid a race condition
I've run the frontend tests thirce locally, so I think it should be okay, but they might be flaky @bansalnitish